### PR TITLE
Add Enum Support

### PR DESCRIPTION
**Pull request by matt-attack**
Also allows preserving declaration order of parameters instead of having them displayed in alphabetical order.

### CODE REVIEW

File: swri_roscpp/include/swri_roscpp/dynamic_parameters.h [hunk 1]
```diff
@@ -392,6 +426,34 @@ namespace swri
         desc.level = 0;
         desc.description = param->second.description;
         desc.edit_method = "";
+
+        // If this is an enum, lets make the edit method string
+        for (int i = 0; i < param->second.enums.size(); i++)
```
Comment (pjreed) [inline swri_roscpp/include/swri_roscpp/dynamic_parameters.h]: This looks like it is shadowing the `i` parameter from the outer loop, is that intentional?  And is there a reason to not use for-each loops here?

File: swri_roscpp/include/swri_roscpp/dynamic_parameters.h [hunk 2]
```diff
@@ -59,6 +59,7 @@ namespace swri
     Type type;
     std::string name;
     std::string description;
+    std::vector<std::pair<std::string, int>> enums;
```
Comment (pjreed) [inline swri_roscpp/include/swri_roscpp/dynamic_parameters.h]: Not having a space between the `>`'s will break some versions of gcc, although I'm not sure if that's an issue on Ubuntu 16.04 / 18.04.
```suggestion
    std::vector<std::pair<std::string, int> > enums;
```

File: swri_roscpp/include/swri_roscpp/dynamic_parameters.h [hunk 3]
```diff
@@ -392,6 +426,34 @@ namespace swri
         desc.level = 0;
         desc.description = param->second.description;
         desc.edit_method = "";
+
+        // If this is an enum, lets make the edit method string
+        for (int i = 0; i < param->second.enums.size(); i++)
+        {
+          if (i == 0)
+          {
+            desc.edit_method += "{'enum_description': '" + escapeString(desc.description) + "'";
```
Comment (pjreed) [inline swri_roscpp/include/swri_roscpp/dynamic_parameters.h]: Assembling YAML strings manually seems dangerous, especially writing our own method to escape strings.  Why not use yaml-cpp?

File: swri_roscpp/include/swri_roscpp/dynamic_parameters.h [hunk 4]
```diff
@@ -405,6 +467,20 @@ namespace swri
 
       set_service_ = nh_.advertiseService("set_parameters",
             &DynamicParameters::setConfigCallback, this);
+
+      ordered_params_.clear();// to save memory
+    }
+
+    void addEnums(const std::string& param, const std::vector<std::pair<std::string, int>>& enums)
```
Comment (pjreed) [inline swri_roscpp/include/swri_roscpp/dynamic_parameters.h]:
```suggestion
    void addEnums(const std::string& param, const std::vector<std::pair<std::string, int> >& enums)
```

File: swri_roscpp/include/swri_roscpp/dynamic_parameters.h (hunk 1 above)
Comment (matt-attack) [reply swri_roscpp/include/swri_roscpp/dynamic_parameters.h]: Ooops. Should not do that.

I wanted an index so writing out the loop manually is usually the most concise way of doing that.

Comment (matt-attack) [reply swri_roscpp/include/swri_roscpp/dynamic_parameters.h]: Also, it seems some people are requesting this to support C++03 which doesnt allow those.

File: swri_roscpp/include/swri_roscpp/dynamic_parameters.h (hunk 3 above)
Comment (matt-attack) [reply swri_roscpp/include/swri_roscpp/dynamic_parameters.h]: Was going for the most prudent solution and didn't want to add extra dependencies.

**Comment by HoomanBean** [comment]
When is this gonna get merged. I need it.

**Pull request merged**